### PR TITLE
Add sharding tests to multigpu-test.sh and fix custom operator decorator

### PR DESCRIPTION
**Pull request by pritamdamania87**
1. Enabled multigpu tests.
2. Fixed failing multigpu tests.
3. Fixed custom operator decorator to be first preference in operator dispatch.

### CODE REVIEW

**Comment by suo** [comment]
ci-all doesn't do anything, apply the label ciflow/trunk` to run all trunk test

**Comment by suo** [comment]
cc @janeyx99 sounds like something for flaky reruns?

**Comment by pritamdamania87** [comment]
@suo I see, so is this PR safe to land or did some of the multigpu tests not run due to this error?

**Comment by suo** [comment]
I don't know, I would defer to @janeyx99 about whether this is safe

**Comment by janeyx99** [comment]
@pritamdamania87 Since the test could be spawned multiple times + get multiple entries (due to the spawning code or flakiness rerun detection), we manually keep track of a list of tests that are spawned multiple times here for test time tracking purposes. Looking at the logs, `test_custom_sharding_spec` is run twice.
https://github.com/pytorch/pytorch/blob/master/tools/stats/print_test_stats.py#L721

I am writing a change right now that would move this step to another workflow entirely so it shouldn't show up on your PR anymore, but if you don't want to wait for that to land, feel free to add the test file into the MULTITEST.

File: torch/distributed/_shard/sharded_tensor/api.py
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/api.py:790]: what's the relationship between a custom sharded op and a custom sharding_spec op? What if user passed in a custom sharding spec and the custom sharding specs have some ops in there, should a custom sharded op overriding the custom sharding spec op?

My guess is we should override, and this means user facing level, we have two(order matters): 1. custom sharded op 2. custom sharding spec op, we should probably document these two type of registrations to the user. 

internally we have: 1. custom sharded op 2. custom sharding spec op 3. default sharded op (order matters).

**Review by wanchaol — APPROVED** [review]
looks good to me, have some questions on the priority of different types op registration.

**Review by wanchaol — COMMENTED** [review]
oh btw, it's missing the content of `Add sharding tests to multigpu-test.sh`?

File: torch/distributed/_shard/sharded_tensor/api.py
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/api.py:790]: Does this mean we are going to allow users to use their own customized op? Is there already a use case here?

**Review by fduwjj — APPROVED** [review]

**Comment by pritamdamania87** [comment]
> oh btw, it's missing the content of Add sharding tests to multigpu-test.sh?

Ah my bad fixed it.

File: torch/distributed/_shard/sharded_tensor/api.py
Comment (pritamdamania87) [reply torch/distributed/_shard/sharded_tensor/api.py:790]: > Does this mean we are going to allow users to use their own customized op? Is there already a use case here?

Yes, this was actually the reason to add `sharded_op_impl` decorator.

> internally we have: 1. custom sharded op 2. custom sharding spec op 3. default sharded op (order matters).

Yeah, let me add documentation that this is the order in which we allow the custom ops.

**Comment by pritamdamania87** [comment]
@suo @janeyx99 Looks like for the latest update to the PR, the multigpu tests are not being run. I'm not sure what is the reason for this? I tried removing and adding back both `ciflow/trunk` and `ciflow/all` tags and it still doesn't seem to run the multigpu tests.

**Comment by janeyx99** [comment]
@pritamdamania87 Wait I haven't landed my change to move out print_test_stats yet, so the multigpu tests will still fail on trunk (as I see you haven't added the file for test_custom_sharding_spec into MULTITESTS) :/

**Pull request closed without merging**
